src/core/ext/filters/client_channel/lb_policy/xds/xds_routing.cc

```diff
@@ -0,0 +1,833 @@
+          "value should be of type object"));
+      return error_list;
+    }
+    // Parse service
+    auto it = json.object_value().find("service");
+    if (it != json.object_value().end()) {
+      if (it->second.type() != Json::Type::STRING) {
+        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+            "field:service error: should be string"));
+      } else {
+        route_config->service = it->second.string_value();
+      }
+    }
+    // Parse method
+    it = json.object_value().find("method");
+    if (it != json.object_value().end()) {
+      if (it->second.type() != Json::Type::STRING) {
+        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+            "field:method error: should be string"));
+      } else {
+        route_config->method = it->second.string_value();
+      }
+    }
+    if (route_config->service.empty() && !route_config->method.empty()) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "service is empty when method is not"));
+    }
+    return error_list;
+  }
+
+  static std::vector<grpc_error*> ParseRoute(
+      const Json& json, const XdsRoutingLbConfig::ActionMap& action_map,
+      XdsRoutingLbConfig::Route* route,
+      std::set<std::string /*action_name*/>* actions_to_be_used) {
+    std::vector<grpc_error*> error_list;
+    if (json.type() != Json::Type::OBJECT) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "value should be of type object"));
+      return error_list;
+    }
+    // Parse MethodName.
+    auto it = json.object_value().find("methodName");
+    if (it == json.object_value().end()) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "field:methodName error:required field missing"));
+    } else {
+      std::vector<grpc_error*> method_name_errors =
+          ParseMethodName(it->second, &route->matcher);
+      if (!method_name_errors.empty()) {
+        error_list.push_back(GRPC_ERROR_CREATE_FROM_VECTOR(
+            "field:methodName", &method_name_errors));
+      }
+    }
+    // Parse action.
+    it = json.object_value().find("action");
+    if (it == json.object_value().end()) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "field:action error:required field missing"));
+    } else if (it->second.type() != Json::Type::STRING) {
+      error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+          "field:action error:should be of type string"));
+    } else {
+      route->action = it->second.string_value();
+      if (route->action.empty()) {
+        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
+            "field:action error:cannot be empty"));
+      } else {
+        // Validate action exists and mark it as used.
+        if (action_map.find(route->action) == action_map.end()) {
+          error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
```

Comment:
I came across this code and found using GRPC_ERROR_CREATE_FROM_STATIC_STRING here is probably a bug. Should probably be (GRPC_ERROR-CREATE_FROM_COPIED_STRING)

Reply:
Good catch, Muxi!  Donna, can you please fix this in a separate PR?  Thanks!